lifetimes/models/__init__.py

```diff
@@ -0,0 +1,181 @@
+
+from abc import ABC, abstractmethod
+from copy import deepcopy
+import warnings
+import json
+import psutil
+from typing import Dict, List, Iterable, TypeVar, Generic
+
+import numpy as np
+import pandas as pd
+
+import pymc as pm
+import arviz as az
+import aesara.tensor as at
+
+from ..utils import _check_inputs
+
+
+SELF = TypeVar("SELF")
+
+class BaseModel(ABC, Generic[SELF]):
+
+    # This attribute must be defined in subclasses.
+    remove_hypers: list
+
+    @abstractmethod
+    def _model() -> None:
+        """ pymc model defining priors for model parameters and calling _loglike in Potential()."""
+        pass
+
+    @abstractmethod
+    def _loglike() -> None:
+        """ Log-likelihood function for randomly drawn individual from customer population. Must be constructed from aesara tensors. """
+        pass
+    
+    @abstractmethod
+    def predict() -> None:
+        pass
+
+    def __repr__(self) -> str:
+        """Representation of BTYD model object."""
+        classname = self.__class__.__name__
+        try:
+            row_str = f"estimated with {self.frequency.shape[0]} customers."
+        except AttributeError:
+            row_str = ""
+
+        try:
+            param_keys = [key.split(f'{classname}::')[1] for key in list(self.param_dict.get('data_vars').keys())]
+            param_vals = np.around(self._unload_params(),decimals=1)
+            param_str = str(dict(zip(param_keys, param_vals)))
+            return f"<btyd.{classname}: Parameters {param_str} {row_str}>"
+        except AttributeError:
+            return f"<btyd.{classname}>"
+    
+    def fit(self, rfm_df: pd.DataFrame, tune: int = 1200, draws: int = 1200) -> SELF:
+        """
+        Fit a custom pymc model with parameter prior definitions to observed RFM data.
+
+        Parameters
+        ----------
+        rfm_df: pandas.DataFrame
+            Pandas dataframe containing customer ids, frequency, recency, T and monetary value columns.
+        tune: int
+            Number of beginning 'burn-in' samples for posterior parameter distribution convergence. These are discarded after model is fit.
+        draws: int
+            Number of samples from posterior parameter distrutions after tune period. These are retained for model usage.
+        chain: int
+            Number of sampling chains used for inference. It is best to set this to the number of CPU cores.
+        
+        Returns
+        -------
+        param_posteriors : dict
+            Posterior distributions of model parameters and other inference metadata.
+        
+        """
+
+        self.frequency, self.recency, self.T, self.monetary_value, _ = self._dataframe_parser(rfm_df)
+
+        #self.model = self._model()
+
+        with self._model():
+            self.idata = pm.sample(
+                tune=tune,
+                draws=draws,
+                chains=4,
+                cores=len(psutil.Process().cpu_affinity()), 
+                target_accept=0.95,
```

Comment:
Can we add a `sample_kwargs` as an input for the function? Sometimes one would like to modify them, for example the `target_accept` value.

Reply:
That's a great suggestion; I'll create an issue for it.